org.eclipse.tm4e.core/src/main/java/org/eclipse/tm4e/core/model/AbstractModelLines.java

```diff
@@ -37,7 +37,7 @@ static final class ModelLine {
 		List<TMToken> tokens = Collections.emptyList();
 	}
 
-	private final List<ModelLine> list = new LinkedList<>();
+	private final List<ModelLine> list = new ArrayList<>();
```

Comment:
This is a nice starting point.
I suggest to go a bit further and try to work based on addAll(int idx, Collection elements), and list.subList().clear();

Two examples:

```
protected void addLines(final int lineIndex, final int count) {
		if (count < 1)
			return;
		// prepare temporary list;
		var newLines =    new ArrayList<>(count)
		for (int i = 0; i < count; i++) {
			newLines.add(new ModelLine());
		}
		synchronized (list) {
			final var firstLine = getOrNull(lineIndex);
			list.addAll(lineIndex, newLines);
			if (firstLine != null) {
				list.get(lineIndex).startState = firstLine.startState;
			}
			updateLine(lineIndex);
		}
	}
```

similarly `removeLines` could benefit from `list.sublist(lineIndex, lineIndex + count).clear()`

Reply:
Likewise there is potential for optimizations in replaceLines via `subList().replaceAll(..)` followed by an addAll or clear depending on the requested changes.

Reply:
> > took 29 seconds on my machine before this PR and takes 20 seconds with this PR
> 
> Could you please provide updated numbers? Because to be honest, 20 seconds is still quite slow.

I shaved of a few more seconds by adding a token caching on the hot path but I don't see more low hanging fruits that will have a noticeable. Because of it's async nature profiling is a bit difficult. Most of the time seems now to be spend in the joni library for TextMate regex searches and in eclipse UI classes doing the rendering. I see one more possible optimization but I am not sure how/if this will result in a noticeable performance improvement https://github.com/eclipse/tm4e/issues/38